tests/PHPStan/Rules/Arrays/data/invalid-key-array-dim-fetch.php

```diff
@@ -35,3 +35,14 @@
 foreach ($array as $i => $val) {
 	echo $array[$i];
 }
+
+/** @var int|object $maybeInt */
+$maybeInt = null;
+
+/** @var string|array $maybeString */
+$maybeString = [];
+echo $maybeString[$maybeInt];
+
+/** @var mixed $mixed */
+$mixed = null;
+echo $mixed[$maybeInt];
```

Comment:
the first one was reported, the second one not. and the `instanceof MixedType` makes sure to keep that behaviour..

Reply:
This one should be reported on level 9 though. We can fix that.

Reply:
Alright, so the problem is that array dim fetch dim type other than `int|string` might be valid for things that aren't arrays. I think that simply `$varType->isArray()->no()` is correct here. If it might be an array, let's complain about the possibly invalid key.

The example in https://phpstan.org/r/bc4b663d-fa12-4037-aed2-e0ed6d2bb53c is about the `$mixed` part in `$mixed[$maybeInt]`, that rule is gonna complain about `$mixed[$int]` too.

I think that in case of `$mixed[$maybeInt]`, both rules should complain.

So the `$varType instanceof MixedType` is totally unnecessary and also possibly incorrect for subtractions. As we're learning again and again, `instanceof *Type` is always wrong :)

What the problem here is right now is that InvalidKeyInArrayDimFetchRule doesn't use RuleLevelHelper so it's gonna complain about everything right from the start. We should employ `RuleLevelHelper::findTypeToCheck()` here too.

Reply:
ok, that makes sense! I added another commit that introduces usage of it, but I have no experience with it. So I'm not sure if e.g. the `unionTypeCriteriaCallback` is right and also not sure if more tests need to be added with different RuleLevelHelper configuration for example

Reply:
The usage isn't right, and you can test that :) Some individual level 9 bugs are usually tested in tests that have `private bool $checExplicitMixed` property, but if I implement something from scratch, it belongs to LevelsIntegrationTest - just write the `.php` file, add it to `dataTopics` and run the test, it will save the current JSONs for you. And you should probably comment out the other topics meanwhile otherwise it's gonna run for a long time :)

Reply:
it looks like the `stringOffsetAccess` `LevelsIntegrationTest` still complains on level 7. I must be missing something then hmm

Reply:
The pattern for using `findTypeToCheck is`:

```
		$typeResult = $this->ruleLevelHelper->findTypeToCheck(
			$scope,
			$node->expr,
			'', // unknown class error message pattern
			static function (Type $type) use ($castTypeCallback): bool {
				// check something on $type, return false if it's an error
			},
		);
		$type = $typeResult->getType();
		if ($type instanceof ErrorType) { // this part is always the same, some rules report the unknown class reports here
			return [];
		}

		// check the same thing on $type here, report errors if it's false
```

You can check for example InvalidCastRule.

Reply:
ah I think this is related to `findTypeToCheck()` returning `ErrorType`. I'll look closely later

Reply:
we wrote almost at the same time and github didn't show me your comments, so it looks like we kind of ignored each other, weird.
anyways, I adapted it to handle the ErrorType and kept the code that aborts if it isn't an array. it now behaves exactly like on 1.9.x, meaning it ignores mixed basically. I couldn't figure out easily how to let it report for StrictMixed, but I'm not sure if it makes sense tbh. It looks like this rule should only work on things that can be arrays